File: 백준/Silver/25185. 카드 뽑기/카드 뽑기.py
import sys

input = sys.stdin.readline

T = int(input().strip())
results = []

for _ in range(T):
    emotion = ":("

    suit_map = {}

    cards = input().strip().split()
    first_count_same = 0
    second_count_same = 0
    third_count_same = 0

    for i, card in enumerate(cards):
        num = int(card[:-1])
        suit = card[-1]
        if suit not in suit_map:
            suit_map[suit] = set()
        suit_map[suit].add(num)

        if card == cards[0]:
            first_count_same += 1
        if card == cards[1]:
            second_count_same += 1
        if card == cards[2]:
            third_count_same += 1

    for numbers in suit_map.values():
        if len(numbers) >= 3:
            sorted_numbers = sorted(numbers)
            for i in range(len(sorted_numbers) - 2):
                if (sorted_numbers[i] + 1 == sorted_numbers[i + 1] and
                        sorted_numbers[i] + 2 == sorted_numbers[i + 2]):
                    emotion = ":)"
                    break

    if first_count_same > 2 or second_count_same > 2:
        emotion = ":)"
    if first_count_same == 2 and second_count_same == 2 and third_count_same == 2:
        emotion = ":)"

    results.append(emotion)

sys.stdout.write("\n".join(results) + "\n")

File: 백준/Silver/25185. 카드 뽑기/카드 뽑기.java
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int T = Integer.parseInt(br.readLine().trim());
        StringBuilder result = new StringBuilder();
        for(int t = 0; t < T; t++){
            String emotion = ":(";

            Map<Character, Set<Integer>> suitMap = new HashMap<>();

            String[] cards = br.readLine().trim().split(" ");
            int firstCountSame = 0;
            int secondCountSame = 0;
            int thirdCountSame = 0;

            for(String card : cards){
                int num = Integer.parseInt(card.substring(0, card.length() - 1));
                char suit = card.charAt(card.length() - 1);
                suitMap.computeIfAbsent(suit, k -> new HashSet<>()).add(num);

                if(card.equals(cards[0])) firstCountSame++;
                if(card.equals(cards[1])) secondCountSame++;
                if(card.equals(cards[2])) thirdCountSame++;
            }

            for (Set<Integer> numbers : suitMap.values()) {
                if (numbers.size() >= 3) {
                    List<Integer> sortedNumbers = new ArrayList<>(numbers);
                    Collections.sort(sortedNumbers);
                    for (int i = 0; i < sortedNumbers.size() - 2; i++) {
                        if (sortedNumbers.get(i) + 1 == sortedNumbers.get(i + 1) &&
                                sortedNumbers.get(i) + 2 == sortedNumbers.get(i + 2)) {
                            emotion = ":)";
                            break;
                        }
                    }
                }
            }

            if(firstCountSame > 2 || secondCountSame > 2) emotion = ":)";
            if(firstCountSame == 2 && secondCountSame == 2 && thirdCountSame == 2) emotion = ":)";

            result.append(emotion).append('\n');
        }

        bw.write(String.valueOf(result));
        bw.flush();
        br.close();
        bw.close();
    }
}